include/igraph_decls.h

```diff
@@ -13,7 +13,7 @@
  * Reference: https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros */
 #undef DECLDIR
 #if defined (_WIN32)
-    #if defined (__CYGWIN__)
+    #if defined (__GNUC__)
```

Comment:
Can you explain why this is necessary? MinGW supports `__decldir`

Reply:
Building on the question of @szhorvat: how does MinGW work when building a shared library? From [this](http://www.mingw.org/wiki/sampledll) page it seems that it might be possible to build DLLs without using `declspec` declarations, but that would require using `-no-undefined` and `--enable-runtime-pseudo-reloc` options. Can you confirm this is the case?

If MinGW does not require `declspec` declarations for a shared library, that would simplify it a lot. Building a shared library under Windows was not yet supported, and it turned out there were quite some things that needed to be corrected before it built correctly (see PR https://github.com/igraph/igraph/pull/1467 for some of those things). Essentially, on the `develop` branch, we now have a working build system on Windows (using CMake), including a fully functional shared library (both on MSVC and MinGW). However, this is only for the next major release (0.9.0). If `declspec` declarations are required on MinGW, I'm afraid it will be difficult to support building a shared library on MinGW for the 0.8.x series.

We actually do build the library also on MinGW in the CI, see [this](https://ci.appveyor.com/project/ntamas/igraph/builds/36493659/job/n0fcy5deax7169cp) build. That build does passes without this modifications. Locally on my machine, I also need to make some modifications, so I am not sure why this is actually building correctly on the CI.

Reply:
Sorry for the late and only brief reply to the two messages above.  Unfortunately, I am really tied up to the neck right now.

* While MINGW does support `_decldir`, I bumped into a link error for MINGW64 which caused me to look into this.
* I'll try to look into [this](http://www.mingw.org/wiki/sampledll), but please bear in mind that if `autoconf` and `libtool` governs the build process it does become not so easy. [GCC information](https://gcc.gnu.org/onlinedocs/gcc-10.2.0/gcc/Microsoft-Windows-Function-Attributes.html#Microsoft-Windows-Function-Attributes) on `dllimport` and `dllexport`.  
* Using autoconf/libtool, how can static and shared libraries be built? See `DLL_EXPORT`  [here](https://www.gnu.org/software/libtool/manual/libtool.html#Windows-DLLs).
* Note that appveyor (with `MSVC_VERSION=NONE`) calls MSYS2's gcc which is CYGWIN-like, i.e. it does **not** call the MINGW32 or MINGW64 gcc which would be what you might be wanting, too. To verify watch out, e.g., for the target prefix  `x86_64-pc-msys` of the used `gcc` in `config.log`. 
* I hope to get some time to look at the develop branch using `cmake` for MSVC and MinGW.  While I recently spent some time with handling `autoconf` and friends, `cmake` runs a lot more quickly, is more flexible and updated on a regular basis. 

I'll try to put together an easy CI for the various gcc's MSYS2, MINGW32, MINGW64 which might be helpful ... hopefully shortly.

Reply:
>     * While MINGW does support `_decldir`, I bumped into a link error for MINGW64 which caused me to look into this.

The link error can be observed [here](https://github.com/jannick0/igraph/runs/1470659590?check_suite_focus=true#step:4:20202), a log of a quickly composed github workflow for MINGW32 and MINGW64.

Reply:
> I'll try to put together an easy CI for the various gcc's MSYS2, MINGW32, MINGW64 which might be helpful ... hopefully shortly.

#1558 - CI runs for MINGW32 and MINGW64 which currently fail (see my previous comment).

Reply:
Yes, this should be fixed somehow, preferably in the simplest possible way for 0.8.x.

Now all work is being done on the `develop` branch and we use CMake for the build system. [See draft build instructions.](https://igraph.discourse.group/t/draft-document-for-installation-with-cmake/504) Now we build either a static library or a DLL, but not both at the same time. This makes thing simpler. When _using_ an igraph built as a static library on Windows, it is necessary to define `IGRAPH_STATIC` before including any headers. This switches the `DECLDIR` to empty. Otherwise it's "dllimport". The entire decldir thing _should_ work with MinGW as well.

If you have time at some point, it would be really great if you could help set up a CI workflow for MinGW, either on Appveyor, which we already use, or some other CI service. We should make sure that what happened here (broken build with MinGW) won't happen again. This is not urgent since the 0.9 release is quite some time off.

Reply:
> Yes, this should be fixed somehow, preferably in the simplest possible way for 0.8.x.

It could well be that gcc complains about different function declaration and and definition qualifiers, i.e. probably `DECLDIR` should be consistently used in the .h and the .c files.  ... just guessing, not checked yet.     

> [...] When _using_ an igraph built as a static library on Windows, it is necessary to define `IGRAPH_STATIC` before including any headers. This switches the `DECLDIR` to empty.

Why not putting `-DIGRAPH_STATIC` into `igraph.pc` for a static library on Windows?  Alternatively, to have simultaneously static and dynamic libraries, why not shipping `igraph-config.sh` which handles this stuff, assuming that `cmake` does not have a hard time to build both.  
 
> If you have time at some point, it would be really great if you could help set up a CI workflow for MinGW, either on Appveyor, which we already use, or some other CI service. We should make sure that what happened here (broken build with MinGW) won't happen again. This is not urgent since the 0.9 release is quite some time off.

Ok, I'll try to look into that.  Let's do this first to have CI MINGW test runs before tackling the dllimport issue.

Reply:
> > If you have time at some point, it would be really great if you could help set up a CI workflow for MinGW, either on Appveyor, which we already use, or some other CI service. We should make sure that what happened here (broken build with MinGW) won't happen again. This is not urgent since the 0.9 release is quite some time off.
> 
> Ok, I'll try to look into that. Let's do this first to have CI MINGW test runs before tackling the dllimport issue.

#1559 - First version of github workflow ... tests fail due to link issue with static igraph library.